opta/utils/ref.py

```diff
@@ -0,0 +1,166 @@
+import functools
+import re
+from collections import abc
+from typing import Any, Iterable, Iterator, Sequence, Tuple, TypeVar, Union, overload
+
+from opta.utils.yaml import register_yaml_class
+
+_PART_REGEX = re.compile(r"[a-z]([a-z0-9_]*[a-z0-9])?", re.IGNORECASE)
```

Comment:
Is there a reason why a string ending with a underscore is not acceptable here?

Reply:
Not really, but there probably shouldn't be any keys ending in underscores. We can allow it if we want.